nucleus/v1/types.go

```diff
@@ -0,0 +1,178 @@
+
+// HubCoreList is a collection of nucleus hub.
+type HubCoreList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#types-kinds
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Items is a list of nucleus hub.
+	Items []HubCore `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// RegistrationHubSpec represents the desired configuration of registration controller installed on hub.
+type RegistrationHubSpec struct {
+	CommonComponentConfig `json:",inline" protobuf:"bytes,1,opt,name=commonComponentConfig"`
+}
+
+// CommonComponentConfig represents the common configuration of the component.
+type CommonComponentConfig struct {
+	// Image is the image name of the component. If not set, default imagee is used.
+	// +optional
+	Image string `json:"image,omitempty" protobuf:"bytes,1,opt,name=image"`
+}
+
+// WorkAgentSpeec represeents the desired configuration of work agent.
+type WorkAgentSpeec struct {
+	CommonComponentConfig `json:",inline" protobuf:"bytes,1,opt,name=commonComponentConfig"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster
+
+// AgentCore represents a deployment of nucleus agent core on spoke cluster.
+type AgentCore struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec represents the desired deployment configuratioin of nucleus agent.
+	Spec AgentCoreSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status represents the current status of nucleus agent.
+	Status AgentCoreStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// AgentCoreSpec represents the desired deployment configuratioin of nucleus agent.
+type AgentCoreSpec struct {
+	// Namespace is the namespace to deploy the agent.
+	// The namespace must have a prefix of "open-cluster-management-", and if it is not set,
+	// the namespace of "open-cluster-management" is used to deploy agent.
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,1,opt,name=namespace"`
+
+	// Registration represents the desired configuration of registration agent.
+	// +required
+	Registration RegistrationAgentSpec `json:"registration" protobuf:"bytes,2,opt,name=registration"`
+
+	// Work represents the desired configuration of work agent.
+	// +optional
+	Work WorkAgentSpeec `json:"work,omitempty" protobuf:"bytes,3,opt,name=work"`
+
+	// BootstrapKubeconfigSecret is the kubeconfig of hub to bootstrap the process that
+	// agent joins hub.
+	// +required
+	BootstrapKubeconfigSecret string `json:"bootsrapKubeconfigSecret" protobuf:"bytes,4,opt,name=bootstrapKubeconfigSecret"`
```

Comment:
I don't think we need to configure this.  I think we can simply require it to be in a fixed location for our operator.  Say `bootstrap-hub-kubeconfig`